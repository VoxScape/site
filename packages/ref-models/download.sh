#!/usr/bin/env bash
set -uex
cd $(dirname "$0")

# $1: zip URL to download
# $2: local filename
fetch-models () {
  if [[ ! -f "$2" ]]; then
    wget "$1" -O "$2" --continue
  fi
  unzip -o "$2"
}

# => voxel-model-master/
fetch-models https://github.com/ephtracy/voxel-model/archive/refs/heads/master.zip ephtracy-voxel-model-master.zip &
# => voxel-model-master/
fetch-models https://github.com/lquesada/voxel-3d-models/archive/refs/heads/master.zip lquesada-voxel-3d-models-master.zip &
fetch-models https://github.com/kluchek/vox-models/archive/refs/heads/master.zip kluchek-vox-models-master.zip &
wait
