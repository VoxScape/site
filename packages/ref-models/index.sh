#!/usr/bin/env bash
cd $(dirname "$0")

find -name '*.vox' | xargs ../vox.ts/bin/voxprobe > index.ndjson 2> index.err
