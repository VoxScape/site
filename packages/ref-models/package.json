{
  "name": "@voxscape/ref-models",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "BSD",
  "devDependencies": {
    "@types/node": "18",
    "@voxscape/vox.ts": "workspace:*"
  },
  "dependencies": {
    "node-find": "^0.7.2",
    "ts-node": "^10.9.1"
  }
}
