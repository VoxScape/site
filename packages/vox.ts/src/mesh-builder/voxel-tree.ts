import { off } from 'process';
import { ParsedVoxFile, VoxelModel, Voxel, VoxelGroup } from '../types/vox-types';

type Axis = 'x' | 'y' | 'z';

const nextAxisLoop = {
  x: 'y',
  y: 'z',
  z: 'x',
} as const;

const MIN_THRESHOLD = 8; // making voxelTree.voxels.length < 8*8*8 voxels

// 3-kdtree
class VoxelTree {
  #voxels: readonly Voxel[];
  readonly leftChild?: VoxelTree;
  readonly rightChild?: VoxelTree;

  constructor(
    voxels: Voxel[],
    readonly threshold: number,
    readonly axis: Axis,
    readonly depth: number,
    readonly offset: number,
    readonly size: number,
  ) {
    const sortedVoxels = voxels.slice(offset, offset + size).sort((v1, v2) => v1[axis] - v2[axis]);
    for (let i = 0; i < size; i++) {
      // must not use splice() here. splice() will stack overflow with large `size`
      voxels[offset + i] = sortedVoxels[i];
    }

    this.#voxels = voxels;

    if (threshold > MIN_THRESHOLD) {
      const nextAxis = nextAxisLoop[axis];
      const nextThreshold = nextAxis === 'x' ? threshold / 2 : threshold;

      const leftSize = sortedVoxels.filter((v) => v[axis] < threshold).length;
      const rightSize = size - leftSize;

      if (leftSize) {
        this.leftChild = new VoxelTree(voxels, nextThreshold, nextAxis, 1 + depth, offset, leftSize);
      }
      if (rightSize) {
        this.rightChild = new VoxelTree(voxels, nextThreshold, nextAxis, 1 + depth, offset + leftSize, rightSize);
      }
    }
  }

  get voxels(): readonly Voxel[] {
    return this.#voxels.slice(this.offset, this.offset + this.size)
  }
  get isLeaf(): boolean {
    return !this.leftChild && !this.rightChild
  }
}

export function buildVoxelTree(voxelModel: VoxelModel): VoxelTree {
  const maxDimension = Math.max(voxelModel.size.x, voxelModel.size.y, voxelModel.size.z);
  const nextPower2 = Math.floor(Math.log2(maxDimension));

  const initialThreshold = 2 ** (nextPower2 - 1);

  return new VoxelTree(voxelModel.voxels.slice(), initialThreshold, 'x', 0, 0, voxelModel.voxels.length);
}
