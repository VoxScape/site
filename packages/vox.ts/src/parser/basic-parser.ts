import type * as VoxTypes from '../types/vox-types';
import { RiffLense } from '../util/riff-lense';
import { getDefaultPalette, readMatt, readRgba, readXyzi, uint32ToColor } from './chunk-reader';

/**
 * parser for "basic" vox format
 */
export function basicParser(
  bytes: ArrayBuffer,
  /** @deprecated prefer to swap yz in mesh builder */ flipYZ = false,
  enableExtension = false,
): VoxTypes.ParsedVoxFile {
  const l = new RiffLense(bytes);

  if (l.asciiAt(0, 4) !== 'VOX ') throw new Error(`expected 'VOX ' at offset=0`);
  if (l.int32At(4) !== 150) throw new Error(`expect 0x150 at offset=4`);

  const mainChunk = l.chunkAt(8);
  if (mainChunk.id !== 'MAIN') throw new Error('expected MAIN chunk');

  const mainChildren = l.childrenChunksIn(mainChunk);

  const models: VoxTypes.VoxelModel[] = [];
  let palette: undefined | VoxTypes.VoxelPalette = undefined;
  const materials: VoxTypes.VoxelMaterial[] = [];
  const warnings: string[] = [];
  for (let chunkIndex = 0; chunkIndex < mainChildren.length; ++chunkIndex) {
    const chunk = mainChildren[chunkIndex];
    if (chunk.id === 'PACK' && chunkIndex === 0) {
      // ignore
    } else if (chunk.id === 'SIZE') {
      const xyzi = mainChildren[chunkIndex + 1];
      if (!(xyzi && xyzi.id === 'XYZI')) {
        throw new Error(`expected 'XYZI' chunk at offset=0x${chunk.end.toString(16)}`);
      }
      models.push(readXyzi(l, chunk, xyzi, flipYZ));
      ++chunkIndex;
    } else if (chunk.id === 'RGBA') {
      palette = readRgba(l, chunk);
    } else if (chunk.id === 'MATT') {
      materials.push(readMatt(l, chunk));
    } else {
      switch (chunk.id) {
        case 'MATL':
        case 'nTRN':
        case 'nGRP':
        case 'nSHP':
        case 'LAYR':
        case 'rOBJ':
        case 'STRING':
        case 'DICT':
        case 'ROTATION':
        case 'NOTE':
        case 'IMAP': {
          const msg = `vox.ts: unsupported ${chunk.id} chunk at offset 0x${chunk.start.toString(16)}`;
          warnings.push(msg);
          break;
        }

        default: {
          warnings.push(`vox.ts: unknown ${chunk.id} chunk at offset 0x${chunk.start.toString(16)}`);
          break;
        }
      }
    }
  }

  return {
    models,
    palette: palette ?? getDefaultPalette(),
    materials,
    warnings,
    enableExtension,
  };
}
