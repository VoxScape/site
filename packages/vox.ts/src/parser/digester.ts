import { ParsedVoxFile, VoxelModel, VoxelModelSize } from '../types/vox-types';

export interface VoxFileDigest {
  path: string;
  warnings: readonly string[];
  models: {
    size: VoxelModelSize;
    numVoxels: number;
  }[];
}

export function digestFile(path: string, f: ParsedVoxFile): VoxFileDigest {
  return {
    path,
    warnings: f.warnings,
    models: f.models.map((m) => ({
      size: m.size,
      numVoxels: m.voxels.length,
    })),
  };
}
