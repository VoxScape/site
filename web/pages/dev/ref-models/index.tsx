import { NextPage } from 'next';
import useConstant from 'use-constant';
import { fetchBaseModelIndex } from '../../../src/components/dev/ref-models';
import { usePromised } from '@jokester/ts-commonutil/lib/react/hook/use-promised';
import { PageMeta } from '../../../src/components/meta/page-meta';
import { RefModelsTable } from '../../../src/components/dev/ref-models-table';

const RefModelsIndexPage: NextPage = () => {
  const indexP = useConstant(() => fetchBaseModelIndex());
  const index = usePromised(indexP);

  return (
    <div>
      <PageMeta title="ref models" />
      {index.fulfilled ? (
        <RefModelsTable
          files={index.value}
          onClick={(m) => {
            window.open(`/dev/ref-models/show?path=${encodeURIComponent(m.path)}`, '_blank');
          }}
        />
      ) : (
        'loading...'
      )}
    </div>
  );
};

export default RefModelsIndexPage;
