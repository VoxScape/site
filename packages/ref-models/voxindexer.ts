import path from 'node:path';
import fsp from 'node:fs/promises';
import * as nodeFind from 'node-find';
import { basicParser } from '@voxscape/vox.ts/src/parser/basic-parser';

async function main(start: string) {
  if (!start) {
    throw new Error(`usage: ${path.basename(__filename)} <start>`);
  }

  const finder = nodeFind.find(nodeFind.and(nodeFind.name('*.vox'), nodeFind.type('f')), { start });

  const index: {
    path: string;
    size: number;
    numModels: number;
    numWarnings: number;
  }[] = [];
  for await (const child of finder) {
    const filename = child.steps && path.join(...child.steps);

    const buffer = await fsp.readFile(filename);

    try {
      const parsed = basicParser(buffer.buffer, false, true);
      index.push({
        path: filename,
        size: buffer.byteLength,
        numModels: parsed.models.length,
        numWarnings: parsed.warnings.length,
      });
    } catch (e) {
      console.error(`error parsing ${filename}`, e);
    }

    console.debug('steps', filename, buffer.byteLength);
  }

  await fsp.writeFile('index.json', JSON.stringify(index, null, 2));
}

setTimeout(() => main(process.argv[2]));
