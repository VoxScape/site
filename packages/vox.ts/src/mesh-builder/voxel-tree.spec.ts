import { chrFoxVox, monu8Vox } from "../__test__/ref-models"
import { basicParser } from "../parser/basic-parser"
import { buildVoxelTree } from "./voxel-tree"

describe('VoxelTree', () => {
    it('builds kdtree from few voxels', () => {
        const parsed = basicParser(chrFoxVox)
        const build = buildVoxelTree(parsed.models[0])
        expect(build).toMatchSnapshot("buildVoxelTree(chrFox.vox)")
    })
    it('builds kdtree from many voxels', () => {
        const parsed = basicParser(monu8Vox)
        const build = buildVoxelTree(parsed.models[0])
        expect(build).toMatchSnapshot("buildVoxelTree(monu8.vox)")
    })

})