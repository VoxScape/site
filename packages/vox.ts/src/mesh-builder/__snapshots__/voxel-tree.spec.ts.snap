// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoxelTree builds kdtree from a lot of voxels: buildVoxelTree(monu0.vox) 1`] = `
VoxelTree {
  "axis": "x",
  "depth": 0,
  "leftChild": undefined,
  "offset": 0,
  "rightChild": VoxelTree {
    "axis": "y",
    "depth": 1,
    "leftChild": VoxelTree {
      "axis": "z",
      "depth": 2,
      "leftChild": undefined,
      "offset": 0,
      "rightChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": undefined,
        "offset": 0,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 0,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": undefined,
            "offset": 0,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 0,
              "rightChild": undefined,
              "size": 248,
              "threshold": 8,
            },
            "size": 248,
            "threshold": 16,
          },
          "size": 248,
          "threshold": 16,
        },
        "size": 248,
        "threshold": 16,
      },
      "size": 248,
      "threshold": 32,
    },
    "offset": 0,
    "rightChild": VoxelTree {
      "axis": "z",
      "depth": 2,
      "leftChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": undefined,
        "offset": 248,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 248,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": undefined,
            "offset": 248,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 248,
              "rightChild": undefined,
              "size": 548,
              "threshold": 8,
            },
            "size": 548,
            "threshold": 16,
          },
          "size": 548,
          "threshold": 16,
        },
        "size": 548,
        "threshold": 16,
      },
      "offset": 248,
      "rightChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": undefined,
        "offset": 796,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 796,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": undefined,
            "offset": 796,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 796,
              "rightChild": undefined,
              "size": 11921,
              "threshold": 8,
            },
            "size": 11921,
            "threshold": 16,
          },
          "size": 11921,
          "threshold": 16,
        },
        "size": 11921,
        "threshold": 16,
      },
      "size": 12469,
      "threshold": 32,
    },
    "size": 12717,
    "threshold": 32,
  },
  "size": 12717,
  "threshold": 32,
}
`;

exports[`VoxelTree builds kdtree from few voxels: buildVoxelTree(chrFox.vox) 1`] = `
VoxelTree {
  "axis": "x",
  "depth": 0,
  "leftChild": undefined,
  "offset": 0,
  "rightChild": undefined,
  "size": 565,
  "threshold": 8,
}
`;

exports[`VoxelTree builds kdtree from quite a lot of voxels: buildVoxelTree(monu8.vox) 1`] = `
VoxelTree {
  "axis": "x",
  "depth": 0,
  "leftChild": VoxelTree {
    "axis": "y",
    "depth": 1,
    "leftChild": VoxelTree {
      "axis": "z",
      "depth": 2,
      "leftChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 0,
              "rightChild": undefined,
              "size": 2366,
              "threshold": 8,
            },
            "offset": 0,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 2366,
              "rightChild": undefined,
              "size": 1352,
              "threshold": 8,
            },
            "size": 3718,
            "threshold": 16,
          },
          "offset": 0,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 3718,
              "rightChild": undefined,
              "size": 2912,
              "threshold": 8,
            },
            "offset": 3718,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 6630,
              "rightChild": undefined,
              "size": 1664,
              "threshold": 8,
            },
            "size": 4576,
            "threshold": 16,
          },
          "size": 8294,
          "threshold": 16,
        },
        "offset": 0,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 8294,
              "rightChild": undefined,
              "size": 2912,
              "threshold": 8,
            },
            "offset": 8294,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 11206,
              "rightChild": undefined,
              "size": 1664,
              "threshold": 8,
            },
            "size": 4576,
            "threshold": 16,
          },
          "offset": 8294,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 12870,
              "rightChild": undefined,
              "size": 3584,
              "threshold": 8,
            },
            "offset": 12870,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 16454,
              "rightChild": undefined,
              "size": 2052,
              "threshold": 8,
            },
            "size": 5636,
            "threshold": 16,
          },
          "size": 10212,
          "threshold": 16,
        },
        "size": 18506,
        "threshold": 16,
      },
      "offset": 0,
      "rightChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": undefined,
        "offset": 18506,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 18506,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": undefined,
            "offset": 18506,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 18506,
              "rightChild": undefined,
              "size": 108,
              "threshold": 8,
            },
            "size": 108,
            "threshold": 16,
          },
          "size": 108,
          "threshold": 16,
        },
        "size": 108,
        "threshold": 16,
      },
      "size": 18614,
      "threshold": 32,
    },
    "offset": 0,
    "rightChild": VoxelTree {
      "axis": "z",
      "depth": 2,
      "leftChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 18614,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 18614,
              "rightChild": undefined,
              "size": 12740,
              "threshold": 8,
            },
            "offset": 18614,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 31354,
              "rightChild": undefined,
              "size": 7280,
              "threshold": 8,
            },
            "size": 20020,
            "threshold": 16,
          },
          "size": 20020,
          "threshold": 16,
        },
        "offset": 18614,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 38634,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 38634,
              "rightChild": undefined,
              "size": 15680,
              "threshold": 8,
            },
            "offset": 38634,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 54314,
              "rightChild": undefined,
              "size": 10463,
              "threshold": 8,
            },
            "size": 26143,
            "threshold": 16,
          },
          "size": 26143,
          "threshold": 16,
        },
        "size": 46163,
        "threshold": 16,
      },
      "offset": 18614,
      "rightChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": undefined,
        "offset": 64777,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 64777,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": undefined,
            "offset": 64777,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 64777,
              "rightChild": undefined,
              "size": 3350,
              "threshold": 8,
            },
            "size": 3350,
            "threshold": 16,
          },
          "size": 3350,
          "threshold": 16,
        },
        "size": 3350,
        "threshold": 16,
      },
      "size": 49513,
      "threshold": 32,
    },
    "size": 68127,
    "threshold": 32,
  },
  "offset": 0,
  "rightChild": VoxelTree {
    "axis": "y",
    "depth": 1,
    "leftChild": VoxelTree {
      "axis": "z",
      "depth": 2,
      "leftChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": undefined,
        "offset": 68127,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 68127,
              "rightChild": undefined,
              "size": 12740,
              "threshold": 8,
            },
            "offset": 68127,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 80867,
              "rightChild": undefined,
              "size": 7280,
              "threshold": 8,
            },
            "size": 20020,
            "threshold": 16,
          },
          "offset": 68127,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 88147,
              "rightChild": undefined,
              "size": 15680,
              "threshold": 8,
            },
            "offset": 88147,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 103827,
              "rightChild": undefined,
              "size": 11583,
              "threshold": 8,
            },
            "size": 27263,
            "threshold": 16,
          },
          "size": 47283,
          "threshold": 16,
        },
        "size": 47283,
        "threshold": 16,
      },
      "offset": 68127,
      "rightChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": undefined,
        "offset": 115410,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 115410,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": undefined,
            "offset": 115410,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 115410,
              "rightChild": undefined,
              "size": 6077,
              "threshold": 8,
            },
            "size": 6077,
            "threshold": 16,
          },
          "size": 6077,
          "threshold": 16,
        },
        "size": 6077,
        "threshold": 16,
      },
      "size": 53360,
      "threshold": 32,
    },
    "offset": 68127,
    "rightChild": VoxelTree {
      "axis": "z",
      "depth": 2,
      "leftChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": undefined,
        "offset": 121487,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 121487,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 121487,
              "rightChild": undefined,
              "size": 68600,
              "threshold": 8,
            },
            "offset": 121487,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 190087,
              "rightChild": undefined,
              "size": 47808,
              "threshold": 8,
            },
            "size": 116408,
            "threshold": 16,
          },
          "size": 116408,
          "threshold": 16,
        },
        "size": 116408,
        "threshold": 16,
      },
      "offset": 121487,
      "rightChild": VoxelTree {
        "axis": "x",
        "depth": 3,
        "leftChild": undefined,
        "offset": 237895,
        "rightChild": VoxelTree {
          "axis": "y",
          "depth": 4,
          "leftChild": undefined,
          "offset": 237895,
          "rightChild": VoxelTree {
            "axis": "z",
            "depth": 5,
            "leftChild": undefined,
            "offset": 237895,
            "rightChild": VoxelTree {
              "axis": "x",
              "depth": 6,
              "leftChild": undefined,
              "offset": 237895,
              "rightChild": undefined,
              "size": 39139,
              "threshold": 8,
            },
            "size": 39139,
            "threshold": 16,
          },
          "size": 39139,
          "threshold": 16,
        },
        "size": 39139,
        "threshold": 16,
      },
      "size": 155547,
      "threshold": 32,
    },
    "size": 208907,
    "threshold": 32,
  },
  "size": 277034,
  "threshold": 32,
}
`;
